### ISSUE TITLE
Cannot $set an int64

### ISSUE THREAD

**Issue by jacqueswaller** [mongodb/mongo-php-driver#199]
Initially with the old mongo driver, you could only save Int32 and the extra bits would get truncated. Eventually a runtime configuration `mongo.native_long` was added which would enable the use of Int64. I can't speak for other platforms, but with my current setup trying to do `[$set => ["TestInt" => 4294967295]]` will save an Int32 in the database with value -1 (so it wraps around instead of truncating). I've tried using the same runtime configuration from the old driver but it does not appear to have any effect with mongodb driver.

Similarly, there does not appear to be a `MongoInt32`/`MongoInt64` equivalent within the `MongoDB\BSON` namespace. I could probably write my own little `MongoInt32`/`MongoInt64` classes in PHP that implement `MongoDB\BSON\Serializable` and `MongoDB\BSON\Unserializable`, but it definitely feels like something that's missing in native.

It's worth noting that _reading_ an Int64 from the database seems to work nicely (at least on 64-bit platforms).

Windows 7, 64-bit
Apache 2.4.18, 64-bit TS
PHP 7.0.2, 64-bit TS
Mongo 3.2.0
MongoDB driver 1.1.1

**Comment by bjori** [comment]
I thought PHP7 on Windows had 64bit integers?

was does the following result in for you on this configuration:
```php
<?php
$array = ['$set' => ["TestInt" => 4294967295]];
var_dump($array);
?>
```

**Comment by jacqueswaller** [comment]
PHP7 on Windows _does_ have 64bit integers. The snippet above results in
```
array(1) { ["$set"]=> array(1) { ["TestInt"]=> int(4294967295) } }
```

The loss of precision happens when interacting with the Mongo driver, in that it treats every number as Int32.

`$inc` suffers from the same problem as `$set`. That is, if I try to `[$inc => ["TestInt" => 4294967295]]` I will actually subtract 1 from TestInt. Further testing has revealed that queries suffer from the same problem: If I were to make a query with a filter like `"TestInt" => ['$lt' => 4294967295]`, I'll actually be looking for documents where TestInt is less than -1.

**Comment by jacqueswaller** [comment]
It seems that I lied when I proclaimed that reading an Int64 from the db is fine. It turns out that it comes back as a string and is liable to have its property name mangled.

With the following test document
```
{ 
    "_id" : ObjectId("569d015eb26caf586174252a"), 
    "TestObj" : {
        "Slots" : NumberLong(-429496729552)
    }
}
```

if I fetch it in php and then var_dump the result, I get:
```
object(stdClass)#13 (2) {
  ["_id"]=>
  object(MongoDB\BSON\ObjectID)#11 (1) {
    ["oid"]=>
    string(24) "569d015eb26caf586174252a"
  }
  ["TestObj"]=>
  object(stdClass)#12 (1) {
    ["Slots0"]=>
    string(13) "-429496729552"
  }
}
```

Notice that "Slots" became "Slots0". The appended characters seem to depend on the combination of property name and integer value. With the right values you can end up with invalid UTF8 :(

**Comment by jmikola** [comment]
I'm tracking both issues in this thread with [PHPC-544](https://jira.mongodb.org/browse/PHPC-544) and will aim to have a fix for this in the next 1.1.x release.

**Comment by jmikola** [comment]
@jacqueswaller: The second issue you're seeing here (property name corruption) is actually independent of the first. I've diagnosed it in [PHPC-592](https://jira.mongodb.org/browse/PHPC-592) but will be able to resolve it concurrently with the original serialization issue.

Offhand, were you at all aware that the int64 was being cast to a string by the driver? We actually [log a warning](https://github.com/mongodb/mongo-php-driver/blob/6aaa45d3802e9fd3309c764c4c5016037a0add4f/phongo_compat.h#L111), but I fear most users may be oblivious to that unless they've enabled logging. Only error/critical log messages get turned into exceptions. We certainly don't want to throw here, as that would interrupt control flow for users, but perhaps we can compromise by emitting a PHP warning in this case.

**Comment by jacqueswaller** [comment]
Indeed, I did notice that it was coming back as a string, despite running 64-bit php/driver. When I write
```print(PHP_INT_SIZE . ", " . PHP_INT_MAX);```
I get
```8, 9223372036854775807```

I haven't done anything special for logging beyond starting mongodb with `mongod.exe -logpath C:\MongoDB\logs\mongo.log`. I don't see any integer overflow warnings there, though. Unless you're talking about a php log of some kind.

**Comment by jmikola** [comment]
> I haven't done anything special for logging beyond starting mongodb with `mongod.exe -logpath C:\MongoDB\logs\mongo.log`. I don't see any integer overflow warnings there, though. Unless you're talking about a php log of some kind.

@jacqueswaller: I was referring to the [`mongodb.debug` INI option](http://php.net/manual/en/mongodb.configuration.php). The C driver (i.e. libmongoc) has its own error reporting, and that INI option allows you to hook into it and capture output. This is far from ideal, as most users aren't going to run their applications with debug logging enabled (assuming they're even aware of the option). Starting in 1.2.0, we hope to raise an exception or PHP warning (see: [PHP-313](https://jira.mongodb.org/browse/PHPC-313)).

Anywho, [1.1.3](https://github.com/mongodb/mongo-php-driver/releases/tag/1.1.3) has just been released and should include the necessary fix for your issue. I'll leave this open until you can confirm that it resolves your problem. Thanks for your patience on this.

**Comment by jacqueswaller** [comment]
Confirmed: the wrap around issue when using large numbers seems to be fixed. Doing a `$set` with 4294967295 will set an Int64 in the db, `$inc` will increment instead of decrementing by 1 and querying `$lte` with 4294967295 gives the results I expect (instead of returning documents where TestInt <= -1). Similarly, if I then `$set` over my TestInt with a small number, it converts it to an Int32 in the db. Basically everything behaves the way I'd expect it to :+1:

The "returns large ints as a string" and "mangled property name" issues also both seem to be fixed (at least with the test object I provided before). var_dump now gives me
```
object(stdClass)#15 (2) {
  ["_id"]=>
  object(MongoDB\BSON\ObjectID)#13 (1) {
    ["oid"]=>
    string(24) "569d015eb26caf586174252a"
  }
  ["MiscBin"]=>
  object(stdClass)#14 (1) {
    ["Slots"]=>
    int(-429496729552)
  }
}
```
as expected.
